README.md
slurpy
======

A python wrapper for SLURM commands.


CHANGES.md
# Change-Log for slurpy
# =====================


## To-Do
## -----
-   Add tests.
-   Change '--cancel' to 'cancel' subcommand.
-   Finish developing '--queue' argument.
-   Make filter arguments accept lists (e.g. for name or state)
-   Add ellipses lines before/after the printed output when using either the `--head` or `--tail` arguments.


## Current
## -------
-   Introduced usage of a `logging.Logger` instance named `log`.  This shouldn't necessarily replace usage of `print` everywhere (for example the actual `slurpy` output should perhaps still just use `print`), but it should be used when optional output will be desired based on verbose usage or debugging.
    -   `slurpy/utils.py` has the log initialization routine.
    -   Introduced the `--DEBUG` flag to produce extensive output for debugging purposes.  Stuff that was previously only printed when `args.verbose` was true now uses the `log.info` level of output.
-   Because `sacct` is weird about the `--starttime` parameter (in particular it shows all jobs which are in *any* state for that starttime... i.e. not specificating having *started* after that time), I've introduced an additional filter called from `sacct._filter_lines()` to actually filter results by the start time.

-   `slurpy/utils.py`
    -   `init_log()` [new-function]
        -   Wrapper to construct a logger object from `zcode.inout` module.  Sets the logging level based on `args` arguments, with the logging-file turned off.
-   `slurpy/sacct.py`
    -   `summary()`
        -   In verbose summary, for jobs in the 'PENDING' state, report the wait times (i.e. time since submission) instead of the run times (zero).
        -   In verbose summary, print the earlier and latest submission times for each state.
        -   If no jobs are in a given state, print a blank verbose line instead of '0.0' times.
    -   `_filter_by_time()` [new-function]
        -   Filter the `sacct` output based on the `Start` parameter (partially filtered by the `--startime` sacct argument.


## v0.3 - 2017/03/31
## -----------------
-   Added `--name` argument, 'sacct' results now filtered by job-name.
    -   Currently uses only a case-sensitive check, using 'in' (i.e. if the filtering parameter is inside the job-name, then it passes the check).
-   Added `--starttime` command to show all jobs after a certain submission time.
-   Added `--sort` argument to sort 'sacct' results by the given field.
    -   New method `sacct._sort_lines()` does this, called from `sacct.sacct_results()`.
-   Added `--id` argument to filter by 'JobID' number.  Also works for canceling jobs.
    -   New method `sacct._filter_by_jobid()` does the filtering by parsing the command-line argument specification string.  This is called by `sacct._filter_lines()`.
-   Added `--head` and `--tail` arguments to print only the first and/or last entries from `sacct`.  These are implemented in `utils.print_lines_dicts()`.
-   Added `--watch` argument to repeat execution every X interval (optionally provided with the argument).  This is implemented in `__main__.main()` using a `time.sleep()` call before rerunning the `main()` routine.  Threading might improve this, but I had trouble with that approach.
    -   Also added an `args.clear` parameter and associated behavior in `utils.print_lines_dicts()`.  Just before printing the output dictionaries, if the code is in 'watch' mode and `args.clear == True`, then the screen will be cleared right before printing output.
-   By default, reformat the times-strings output by sacct (replace 'T' separator with optional alternative character.  Set by parameters in `const.py`, with code in `sacct._parse_sacct_line()`.
-   By default, printed output now only shows a subset of available parameters.  Use '--verbose' to include all of them.
    -   New method `utils._filter_fields_in_lines()` is called from within `utils.print_lines_dicts()`.
-   `slurpy/sacct.py`
    -   `_construct_sacct_command()`
        -   New internal method to construct the sacct command to be called, allow optional arguments, etc.
-   `slurpy/__main__.py`
    -   Added the `_DEFAULT_ARG_START` internal parameter which determines the default value for the `--start` argument.  I set the current value to '-7.0', meaning the default start time is seven days in the past from whenever `slurpy` is run.


## v0.2 - 2017/01/14
## -----------------
-   Added 'CANCELLED' and 'TIMEOUT' states.
-   Added `--verbose` command-line argument to modify results.
-   Added `--summary` command-line argument to summarize all jobs reported by 'sacct'.  Currently returns the counts and optionally the min, max, median elapsed time for each state (if the `--verbose` flag is passed).
-   Added `-p`/`--partition` argument to select a particular partition of the cluster.
-   Added `--cancel` command, to cancel submitted jobs.
    -   When the (hardcoded) `_CANCEL_PROMPT` is set to True, then '__main__' will prompt the user to confirm that they want to cancel the jobs.  If `verbose` is enabled, it will also print the jobs about to be canceled.

-  `setup.py`
    -   Added console entry point to setup script, (and appropriately modified `setup.sh`) to install `slurpy` as a script which is runnable from anywhere via the command line.  Working.
-   `slurpy/__main__.py`
    -   
-   `slurpy/sacct.py`
    -   `summary()` [new-method]
        -   API method to parse 'sacct' command results and summarize succinctly.
        -   Reports number of jobs in each 'state'
        -   If `verbose`, then reports min, max, median elapsed time for each state.
-   `slurpy/utils.py` [new-file]
    -   File for general purpose utility methods.  Currently empty.
    -   `print_lines_dicts()` [new-method]
        -   Method for pretty-printing lines retrieved from 'sacct'.
    -   `prompt_yes_no` [new-method]
        -   Prompt the user for a y/n answer for an arbitrary question, take5n from [Recipe 57708](http://code.activestate.com/recipes/577058/).



slurpy/__main__.py
"""Main entry point for `slurpy` scripts.

Default behavior is to run 'sacct' and print parsed output.
"""
from slurpy import sacct, utils

# Prompt the user to confirm before canceling jobs.
_CANCEL_PROMPT = True
# Default value for the `--start` argument.  Should either be `None`, or a float value that
#    specifies how many days before today to set as the starting value.  For example,
#    If now is `2017-03-28T17:00:00` then the value `-2.5` would set the start time to
#    `2017-03-26T05:00:00`.
_DEFAULT_ARG_START = -7.0


def main(args=None, thread=False):
    if args is None:
        args = _init_argparse()
    log = utils.init_log(args)
    # Attach the log to `args` for convenience
    args.log = log
    log.debug("\nslurpy\n")
    log.debug("RUNNING IN DEBUG MODE\n")
    log.debug("args=\n{}".format(args))

    # Cancel / Kill Jobs
    # ------------------
    if args.cancel:
        log.debug("mode: 'cancel', prompt: '{}'".format(_CANCEL_PROMPT))
        if _CANCEL_PROMPT:
            lines, header = sacct.sacct_results(args)
            # Print the jobs about to be canceled
            if args.verbose:
                log.info("\nCancel would end the following jobs:\n")
                utils.print_lines_dicts(lines, header, args)
                print("")
            prompt = "Are you sure you want to cancel {} jobs?".format(len(lines))
            if not utils.prompt_yes_no(prompt, default='no'):
                log.warning("Exiting.")
                return

        from .import scancel
        log.debug("Running `scancel`.")
        scancel.scancel(args)
        return

    # Information sacct, summary, and squeue operations
    # -------------------------------------------------
    if args.summary:
        log.debug("Running `summary`")
        sacct.summary(args)
    elif args.queue:
        log.debug("Running `squeue`")
        from . import squeue
        squeue.squeue(args)
    else:
        log.debug("Running `sacct`")
        sacct.sacct(args)

    # 'Watch' Output: repeatedly printed
    # ----------------------------------
    # Repeatedly call `main` to keep producing new output
    if (args.watch is not None) and (not thread):
        log.debug("Entering 'watch' mode.")
        import time
        # Start look to keep re-calling this (the `main`) method repeatedly every interval
        while True:
            time.sleep(args.watch)
            # set `thread` to True so that subsequent calls don't start their own loops and timers.
            log.debug("Re-running `main`")
            main(args, thread=True)

    return


def _init_argparse():
    import argparse

    # Determine Default Values
    # ------------------------

    # Default start time
    if _DEFAULT_ARG_START is None:
        def_start = None
    #    Interpret `_DEFAULT_ARG_START` as a number of days in the past
    else:
        import datetime
        dt = datetime.datetime.now() + datetime.timedelta(days=_DEFAULT_ARG_START)
        def_start = dt.strftime("%Y-%m-%dT%H:%M:%S")
        # print("def_start = '{}'".format(def_start))

    # Construct Argument Parser
    # -------------------------
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-v", "--verbose", action="store_true", dest="verbose", default=False,
        help="Extended output.")

    parser.add_argument(
        "--DEBUG", action="store_true", dest="debug", default=False,
        help="Lots of output for debugging purposes only.")

    parser.add_argument(
        "--head", nargs='?', dest="head", default=None, const=10,
        help=("Only print first `head` entries (default: 10)"))

    parser.add_argument(
        "--tail", nargs='?', dest="tail", default=None, const=10,
        help=("Only print last  `tail` entries (default: 10)"))

    parser.add_argument(
        "--state", type=str, dest="state", default=None,
        help="state to filter by (e.g. 'RUNNING', 'COMPLETED')")

    parser.add_argument(
        "--start", type=str, dest="start", default=def_start,
        help=("start time for `sacct` quieries in the format "
              "'YYYY-MM-DD[THH:MM[:SS]]' (e.g. `--start 2017-01-01`)."))

    parser.add_argument(
        "--id", type=str, dest="jobid", default=None, nargs='*',
        help=("JobID number to filter by (e.g. `--id 84513996`, `--id 84513996:84514000`"))

    parser.add_argument(
        "--sort", type=str, dest="sort", default=None,
        help=("Sort 'sacct' results by the given parameter field.  "
              "Use `--sort=-KEY` to reverse-sort by 'KEY'."))

    parser.add_argument(
        "--name", type=str, dest="name", default=None,
        help="job name to filter by")

    parser.add_argument(
        "-s", "--summary", action="store_true", dest="summary", default=False,
        help="Print a summary from the current 'sacct' results.")

    parser.add_argument(
        "-w", "--watch", nargs='?', dest="watch", type=int, default=None, const=4,
        help=("Print continuous output refreshed every given internal in seconds (default: 4)"))

    parser.add_argument(
        "--no-clear", action="store_false", dest="clear", default=True,
        help="Clear the screen before printing results.  NOTE: only applies in 'watch' mode.")

    parser.add_argument(
        "-q", "--queue", action="store_true", dest="queue", default=False,
        help="Print the current 'squeue' results.")

    parser.add_argument(
        "-p", "--partition", nargs='?', const="", default=None, dest="partition",
        help="Target a particular 'Partition' of the cluster.")

    # NOTE: this should be changed to a subcommand
    parser.add_argument(
        "--cancel", action="store_true", dest="cancel", default=False,
        help="Cancel submitted SLURM jobs.")

    args = parser.parse_args()
    return args


if __name__ == "__main__":
    main()
